apps/vmq_ql/src/vmq_ql_query_mgr.erl

```diff
@@ -0,0 +1,226 @@
+    end.
+
+handle_call({fetch, Limit}, From, #state{reply_from=undefined, errors=Errors} = State) ->
+    #state{result=ResultTab} = NewState = ensure_result_tab(State),
+    case results_ready(NewState) of
+        true ->
+            {reply, {ResultTab, filter_errors(Errors)}, fetch_results(Limit, NewState)};
+        false ->
+            {noreply, NewState#state{reply_from={From, Limit}}}
+    end.
+
+handle_cast(cancel, State) ->
+    {stop, normal, State}.
+
+handle_info({'DOWN', _, process, Owner, Reason}, #state{owner=Owner} = State) ->
+    lager:debug("VMQL query owner ~p stopped due to ~p", [Owner, Reason]),
+    {stop, normal, State};
+handle_info({query_error, Node, JobPid, Reason}, #state{waiting=Waiting,
+                                                        errors=Errors} = State) ->
+    NewState = fetch_if_ready(State#state{waiting=lists:delete({Node, JobPid}, Waiting),
+                                          errors=[Reason|Errors]}),
+    {noreply, NewState};
+handle_info({results_ready, Node, JobPid, _NumResults},
+            #state{waiting=Waiting, ready=Ready} = State) ->
+    State1 = State#state{waiting=lists:delete({Node, JobPid}, Waiting),
+                           ready=[{Node, JobPid}|Ready]},
+    State2 = fetch_if_ready(State1),
+    {noreply, State2}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+fetch_if_ready(#state{reply_from=ReplyFrom, result=ResultTab, errors=Errors} = State) ->
+    case results_ready(State) of
+        true when ReplyFrom =/= undefined ->
+            {From, Limit} = ReplyFrom,
+            State1 = fetch_results(Limit, State#state{reply_from=undefined}),
+            gen_server:reply(From, {ResultTab, filter_errors(Errors)}),
+            State1;
+        _ ->
+            State
+    end.
+
+filter_errors([]) -> [];
+filter_errors(Errors) -> lists:usort(Errors).
+
+results_ready(#state{waiting=[],ready=[]}) -> true; %% no query worker alive
+results_ready(#state{waiting=[]}) -> true;
+results_ready(_) -> false.
+
+ensure_result_tab(#state{result=undefined} = State) ->
+    State#state{result=ets:new(?MODULE, [public, ordered_set])};
+ensure_result_tab(State) -> State.
+
+fetch_results(Limit, #state{idx=Idx, result=ResultTab, ready=Ready} = State) ->
+    NewIdx =
+    lists:foldl(
+      fun({_Node, Pid}, AccIdx) ->
+              lists:foldl(
+                fun({RowKey, Row}, AccAccIdx) ->
+                        ets:insert(ResultTab, {{RowKey, AccAccIdx}, Row}),
+                        AccAccIdx + 1
+                end, AccIdx, safe_fetch(Pid, desc, Limit))
+      end, Idx, Ready),
+    State#state{idx=NewIdx}.
+
+safe_fetch(Pid, Order, Limit) ->
+    case is_process_alive(Pid) of
+        true ->
+            vmq_ql_query:fetch(Pid, Order, Limit);
```

Comment:
I don't believe the `safe_fetch` with `is_process_alive` checks is needed as the `vmq_ql_query:fetch/3` handles the `noproc` exception.